connect/mirror/src/test/java/org/apache/kafka/connect/mirror/OffsetSyncStoreTest.java

```diff
@@ -57,52 +58,170 @@ public void testOffsetTranslation() {
+            store.sync(tp, 1500, 11000);
+            assertSparseSyncInvariant(store, tp);
+            assertEquals(OptionalLong.of(-1), store.translateDownstream(null, tp, 1499));
+            assertEquals(OptionalLong.of(11000), store.translateDownstream(null, tp, 1500));
+            assertEquals(OptionalLong.of(11001), store.translateDownstream(null, tp, 2000));
+        }
+    }
+
+    @Test
+    public void testKeepMostDistinctSyncs() {
+        // We should not expire more syncs from the store than necessary;
+        // Each new sync should expire at most two other syncs from the cache
+        long iterations = 1000000;
+        long maxStep = Long.MAX_VALUE / iterations;
+        // Test a variety of steps (corresponding to the offset.lag.max configuration)
+        for (long step = 1; step < maxStep; step = (step << 1) + 1)  {
+            try (FakeOffsetSyncStore store = new FakeOffsetSyncStore()) {
+                int lastCount = 1;
+                store.start();
+                for (long offset = 0; offset <= iterations; offset += step) {
+                    store.sync(tp, offset, offset);
+                    // Invariant A: the latest sync is present
+                    assertEquals(offset, store.syncFor(tp, 0).upstreamOffset());
+                    // Invariant D: the earliest sync is present
+                    assertEquals(0L, store.syncFor(tp, 63).upstreamOffset());
+                    int count = countDistinctStoredSyncs(store, tp);
+                    int diff = count - lastCount;
+                    assertTrue(diff >= -1,
+                            "Store expired too many syncs: " + diff + " after receiving offset " + offset);
+                    lastCount = count;
+                }
+            }
+        }
+    }
+
+    private void assertSparseSync(FakeOffsetSyncStore store, long syncOffset, long previousOffset) {
+        assertEquals(OptionalLong.of(previousOffset == -1 ? previousOffset : previousOffset + 1), store.translateDownstream(null, tp, syncOffset - 1));
+        assertEquals(OptionalLong.of(syncOffset), store.translateDownstream(null, tp, syncOffset));
+        assertEquals(OptionalLong.of(syncOffset + 1), store.translateDownstream(null, tp, syncOffset + 1));
+        assertEquals(OptionalLong.of(syncOffset + 1), store.translateDownstream(null, tp, syncOffset + 2));
+    }
+
+    private int countDistinctStoredSyncs(FakeOffsetSyncStore store, TopicPartition topicPartition) {
+        int count = 1;
+        for (int i = 1; i < OffsetSyncStore.SYNCS_PER_PARTITION; i++) {
+            if (store.syncFor(topicPartition, i - 1) != store.syncFor(topicPartition, i)) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    private void assertSparseSyncInvariant(FakeOffsetSyncStore store, TopicPartition topicPartition) {
+        for (int j = 0; j < OffsetSyncStore.SYNCS_PER_PARTITION; j++) {
+            for (int i = 0; i <= j; i++) {
```

Comment:
Invariants B and C deal exclusively with values of `i` and `j`; shouldn't this be reflected here (especially since the `jUpstream == iUpstream` branch below makes it unnecessary to test the case when `i == j`)?
```suggestion
            for (int i = 0; i < j; i++) {
```